File: extensions/vision-nav/agent/src/altnautica_vision_nav/health.py
"""Heartbeat extras publisher.

The plugin owns a ``navigation`` block on the agent's heartbeat. The
publisher keeps a 1 Hz tick that pushes the latest pipeline snapshot
through ``ctx.telemetry.extend("navigation", payload)``. The payload
shape mirrors the GCS-side ``VisionNavTelemetry`` type and the
``cmd_droneStatus.navigation`` validator on the cloud relay.

Field names are camelCase end-to-end. A previous lite-agent gate-9
finding showed that snake_case drift on the wire silently drops the
entire enriched block at the cloud relay, so this module is the only
authority on the spelling.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Optional

from altnautica_vision_nav.mavlink.comp_status import CompanionState
from altnautica_vision_nav.processors.optical_flow_lk import OpticalFlowResult

log = logging.getLogger(__name__)

TICK_INTERVAL_S = 1.0

# Map the comp-198 companion state to the GCS-facing estimator state.
# The mapping mirrors how the operator reads the card today: when the
# companion is active the OF estimator is steady-state, when it is
# critical the OF samples are being dropped (so the estimator is
# degraded), and when the companion is offline there is no estimator.
_COMPANION_TO_ESTIMATOR_STATE = {
    CompanionState.INACTIVE: "off",
    CompanionState.ACTIVE: "converged",
    CompanionState.CRITICAL: "degraded",
    CompanionState.TERMINATING: "failed",
}


def _scale_source_for_topology(topology: Optional[str]) -> Optional[str]:
    """Derive the flow-scale-source label from the rangefinder topology.

    Today this only emits ``"rangefinder"`` (when a rangefinder of
    any topology is present) or ``None``. Once the rangefinder-free
    estimator lands and the active scale source is known per-sample,
    ``"baro"`` / ``"gps"`` / ``"vision"`` join the value set.
    """

    if topology in ("companion", "fc", "both"):
        return "rangefinder"
    return None


class HealthPublisher:
    """Periodically publish a ``navigation`` block on the agent telemetry bus.

    The OFPipeline calls :meth:`update_from_pipeline` after every emit
    so the snapshot stays fresh. The tick runs at 1 Hz and is
    idempotent: calling :meth:`start` twice is a no-op, and
    :meth:`stop` cancels the task cleanly.
    """

    def __init__(
        self,
        *,
        rangefinder_topology: Optional[str] = None,
        recommended_camera_id: Optional[str] = None,
        mode: Optional[str] = None,
        available_estimators: Optional[list[str]] = None,
    ) -> None:
        self._rangefinder_topology = rangefinder_topology
        self._recommended_camera_id = recommended_camera_id
        self._mode = mode
        self._available_estimators = (
            list(available_estimators) if available_estimators else []
        )
        self._latest_flow_quality: Optional[int] = None
        self._latest_flow_rate_hz: Optional[float] = None
        self._latest_distance_m: Optional[float] = None
        self._companion_state: CompanionState = CompanionState.INACTIVE
        self._task: Optional[asyncio.Task[None]] = None
        self._ctx: object | None = None

    # ------------------------------------------------------------------
    # Public mutation API
    # ------------------------------------------------------------------

    def set_topology(self, topology: Optional[str]) -> None:
        """Set the rangefinder topology label.

        ``topology`` is one of ``"companion"``, ``"fc"``, ``"both"`` or
        ``None``. The cloud relay schema accepts exactly these values.
        """

        self._rangefinder_topology = topology

    def set_recommended_camera_id(self, camera_id: Optional[str]) -> None:
        """Set the camera id surfaced as the GCS recommended device."""

        self._recommended_camera_id = camera_id

    def set_mode(self, mode: Optional[str]) -> None:
        """Set the active estimator-mode label surfaced on the heartbeat.

        Values mirror the config ``mode`` literal (``"off"``,
        ``"optical_flow"`` today; ``"optical_flow_degraded"`` and the
        VIO modes are added in later phases). The GCS mode picker
        renders this as the currently-selected option.
        """

        self._mode = mode

    def set_available_estimators(self, estimators: list[str]) -> None:
        """Set the list of estimator keys this plugin instance can run.

        Mirrors the registry at start-up. The GCS uses this to gate the
        mode-picker options so the operator never sees an estimator the
        agent can't instantiate.
        """

        self._available_estimators = list(estimators)

    def update_companion_state(self, state: CompanionState) -> None:
        """Mirror the comp 198 companion state into the heartbeat block."""

        self._companion_state = state

    def update_from_pipeline(
        self,
        *,
        result: OpticalFlowResult,
        flow_rate_hz: float,
        distance_m: Optional[float],
    ) -> None:
        """Refresh the OF-derived snapshot from one pipeline emit."""

        self._latest_flow_quality = int(result.quality)
        self._latest_flow_rate_hz = float(flow_rate_hz)
        self._latest_distance_m = (
            float(distance_m) if distance_m is not None else None
        )

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start(self, ctx: object) -> None:
        """Begin the 1 Hz publish tick. Idempotent."""

        if self._task is not None and not self._task.done():
            return
        self._ctx = ctx
        self._task = asyncio.create_task(self._tick_loop())

    async def stop(self) -> None:
        """Cancel the publish tick."""

        task = self._task
        self._task = None
        if task is not None:
            task.cancel()
            try:
                await task
            except (asyncio.CancelledError, Exception):  # noqa: BLE001
                pass
        self._ctx = None

    # ------------------------------------------------------------------
    # Snapshot
    # ------------------------------------------------------------------

    def snapshot(self) -> dict:
        """Return the current ``navigation`` payload as a plain dict.

        All keys are camelCase. ``flowDistanceM`` is the only field that
        can be ``None``; the rest carry concrete defaults so the GCS
        renders a stable card even when no frames have flown yet.
        """

        estimator_state = _COMPANION_TO_ESTIMATOR_STATE.get(
            self._companion_state, "off"
        )
        scale_source = _scale_source_for_topology(self._rangefinder_topology)
        return {
            "opticalFlowSupported": True,
            "vioSupported": False,
            "rangefinderTopology": self._rangefinder_topology,
            "recommendedCameraId": self._recommended_camera_id,
            "flowQuality": self._latest_flow_quality,
            "flowRateHz": self._latest_flow_rate_hz,
            "flowDistanceM": self._latest_distance_m,
            "vioState": "absent",
            "vioResetCounter": 0,
            "vioQuality": None,
            "companionState": self._companion_state.name.lower(),
            # Additive fields from the estimator-framework. All
            # optional on the GCS side so an older GCS render path
            # stays correct, and so this snapshot stays a strict
            # superset of the previous shape.
            "mode": self._mode,
            "availableEstimators": list(self._available_estimators),
            "estimatorState": estimator_state,
            "flowScaleSource": scale_source,
        }

    # ------------------------------------------------------------------
    # Internals
    # ------------------------------------------------------------------

    async def _tick_loop(self) -> None:
        try:
            while True:
                await self._emit_once()
                await asyncio.sleep(TICK_INTERVAL_S)
        except asyncio.CancelledError:
            return
        except Exception as exc:  # noqa: BLE001
            log.warning("health publisher tick loop failed: %s", exc)

    async def _emit_once(self) -> None:
        ctx = self._ctx
        if ctx is None:
            return
        telemetry = getattr(ctx, "telemetry", None)
        if telemetry is None:
            return
        extend = getattr(telemetry, "extend", None)
        if extend is None:
            return
        try:
            result = extend("navigation", self.snapshot())
            if asyncio.iscoroutine(result):
                await result
        except Exception as exc:  # noqa: BLE001
            log.warning("health publisher emit failed: %s", exc)

File: extensions/vision-nav/agent/src/health.rs
//! Companion-state machine + the `navigation` heartbeat snapshot.
//!
//! The plugin owns a `navigation` block on the agent's heartbeat,
//! pushed through `ctx.telemetry.extend("navigation", ..)`. The payload
//! shape mirrors the GCS `VisionNavTelemetry` type and the cloud relay
//! `cmd_droneStatus.navigation` validator. Field names are camelCase
//! end-to-end (snake_case drift silently drops the block at the relay).
//!
//! [`CompanionState`] is the comp-198 companion node's state; the
//! degradation ladder transitions it (converged -> ACTIVE; a 2 s
//! degraded/failed streak -> CRITICAL) and it maps to a HEARTBEAT
//! `system_status` and to the GCS-facing estimator state.

use rmpv::Value;

pub use crate::pre_arm::CompanionState;

/// MAVLink `MAV_STATE_*` for the companion HEARTBEAT system_status.
pub fn companion_system_status(state: CompanionState) -> u8 {
    match state {
        CompanionState::Inactive => 3,    // MAV_STATE_STANDBY
        CompanionState::Active => 4,      // MAV_STATE_ACTIVE
        CompanionState::Critical => 6,    // MAV_STATE_CRITICAL
        CompanionState::Terminating => 8, // MAV_STATE_FLIGHT_TERMINATION
    }
}

/// Map the companion state to the GCS-facing estimator-state string.
fn companion_to_estimator_state(state: CompanionState) -> &'static str {
    match state {
        CompanionState::Inactive => "off",
        CompanionState::Active => "converged",
        CompanionState::Critical => "degraded",
        CompanionState::Terminating => "failed",
    }
}

/// The auto-detect summary surfaced verbatim on the heartbeat.
#[derive(Debug, Clone, Default)]
pub struct AutodetectSummary {
    pub suggested_mode: Option<String>,
    pub suggested_mode_reason: Option<String>,
    pub detected_camera_count: Option<i32>,
    pub detected_rangefinder_driver: Option<String>,
}

/// The mutable snapshot the pipeline refreshes each tick. `to_value`
/// serializes the `navigation` block (all keys camelCase).
#[derive(Debug, Clone, Default)]
pub struct HealthSnapshot {
    pub rangefinder_topology: Option<String>,
    pub recommended_camera_id: Option<String>,
    pub mode: Option<String>,
    pub available_estimators: Vec<String>,
    pub companion_state: Option<CompanionState>,
    pub flow_quality: Option<i32>,
    pub flow_rate_hz: Option<f32>,
    pub flow_distance_m: Option<f32>,
    pub flow_scale_source: Option<String>,
    pub estimator_state: Option<String>,
    pub imu_source: Option<String>,
    pub imu_rate_hz: Option<f32>,
    pub camera_imu_sync_offset_ms: Option<f32>,
    pub camera_intrinsics_loaded: bool,
    pub vio_state: Option<String>,
    pub vio_reset_counter: u32,
    pub vio_quality: Option<i32>,
    pub feature_count: Option<i32>,
    pub pre_arm_report: Option<Value>,
    pub autodetect: AutodetectSummary,
}

impl HealthSnapshot {
    /// Serialize as the `navigation` payload. Optional fields ride as
    /// `Nil` so an older GCS render path stays correct and the block is
    /// a strict superset of the previous shape.
    pub fn to_value(&self) -> Value {
        let companion = self.companion_state.unwrap_or(CompanionState::Inactive);
        let estimator_state = self
            .estimator_state
            .clone()
            .unwrap_or_else(|| companion_to_estimator_state(companion).to_string());

        let mut entries: Vec<(Value, Value)> = vec![
            (k("opticalFlowSupported"), Value::Boolean(true)),
            (k("vioSupported"), Value::Boolean(true)),
            (k("rangefinderTopology"), opt_str(&self.rangefinder_topology)),
            (k("recommendedCameraId"), opt_str(&self.recommended_camera_id)),
            (k("flowQuality"), opt_i32(self.flow_quality)),
            (k("flowRateHz"), opt_f32(self.flow_rate_hz)),
            (k("flowDistanceM"), opt_f32(self.flow_distance_m)),
            (k("vioState"), opt_str(&self.vio_state.clone().or(Some("absent".into())))),
            (k("vioResetCounter"), Value::from(self.vio_reset_counter)),
            (k("vioQuality"), opt_i32(self.vio_quality)),
            (k("companionState"), Value::from(companion.as_str())),
            (k("mode"), opt_str(&self.mode)),
            (
                k("availableEstimators"),
                Value::Array(self.available_estimators.iter().map(|s| Value::from(s.as_str())).collect()),
            ),
            (k("estimatorState"), Value::from(estimator_state)),
            (k("flowScaleSource"), opt_str(&self.flow_scale_source)),
            (k("imuSource"), opt_str(&self.imu_source)),
            (k("imuRateHz"), opt_f32(self.imu_rate_hz)),
            (k("cameraImuSyncOffsetMs"), opt_f32(self.camera_imu_sync_offset_ms)),
            (k("cameraIntrinsicsLoaded"), Value::Boolean(self.camera_intrinsics_loaded)),
            (k("estimatorFeatureCount"), opt_i32(self.feature_count)),
            (
                k("preArmReport"),
                self.pre_arm_report.clone().unwrap_or(Value::Nil),
            ),
            (k("suggestedMode"), opt_str(&self.autodetect.suggested_mode)),
            (
                k("suggestedModeReason"),
                opt_str(&self.autodetect.suggested_mode_reason),
            ),
            (k("detectedCameraCount"), opt_i32(self.autodetect.detected_camera_count)),
            (
                k("detectedRangefinderDriver"),
                opt_str(&self.autodetect.detected_rangefinder_driver),
            ),
        ];
        // Keep insertion order stable for tests.
        entries.shrink_to_fit();
        Value::Map(entries)
    }
}

fn k(s: &str) -> Value {
    Value::from(s)
}
fn opt_str(s: &Option<String>) -> Value {
    match s {
        Some(v) => Value::from(v.as_str()),
        None => Value::Nil,
    }
}
fn opt_i32(v: Option<i32>) -> Value {
    match v {
        Some(x) => Value::from(x),
        None => Value::Nil,
    }
}
fn opt_f32(v: Option<f32>) -> Value {
    match v {
        Some(x) => Value::F64(x as f64),
        None => Value::Nil,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn field<'a>(v: &'a Value, key: &str) -> Option<&'a Value> {
        match v {
            Value::Map(e) => e.iter().find(|(k, _)| k.as_str() == Some(key)).map(|(_, val)| val),
            _ => None,
        }
    }

    #[test]
    fn snapshot_is_camelcase_superset() {
        let snap = HealthSnapshot {
            mode: Some("optical_flow".into()),
            companion_state: Some(CompanionState::Active),
            flow_quality: Some(120),
            available_estimators: vec!["off".into(), "optical_flow".into()],
            ..Default::default()
        };
        let v = snap.to_value();
        assert_eq!(field(&v, "mode").unwrap().as_str(), Some("optical_flow"));
        assert_eq!(field(&v, "flowQuality").unwrap().as_i64(), Some(120));
        assert_eq!(field(&v, "companionState").unwrap().as_str(), Some("active"));
        // Active companion maps the GCS estimator state to converged.
        assert_eq!(field(&v, "estimatorState").unwrap().as_str(), Some("converged"));
        // vioSupported is now true (the Rust agent ships VIO modes).
        assert_eq!(field(&v, "vioSupported").unwrap().as_bool(), Some(true));
    }

    #[test]
    fn unset_optional_fields_serialize_nil() {
        let snap = HealthSnapshot::default();
        let v = snap.to_value();
        assert!(matches!(field(&v, "flowDistanceM"), Some(Value::Nil)));
        assert!(matches!(field(&v, "imuSource"), Some(Value::Nil)));
    }

    #[test]
    fn companion_system_status_maps() {
        assert_eq!(companion_system_status(CompanionState::Active), 4);
        assert_eq!(companion_system_status(CompanionState::Critical), 6);
    }
}
